File: java/leetcode/BalanceTree.java
//Leetcode problem 110 Balanced Binary Tree
//Solution written by Xuqiang Fang on 29 March 2018 
import java.util.ArrayList;
import java.util.List;
class TreeNode{
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x){val = x;}
}
class Solution{
	public boolean isBalanced(TreeNode root){
		if(root == null)
			return true;
		if(Math.abs(height(root.left) - height(root.right)) > 1){
			//System.out.println(root.val);
			//System.out.println(height(root.left));
			//System.out.println(height(root.right));
			return false;
		}
		return isBalanced(root.left) && isBalanced(root.right);
	}	
	public void print(TreeNode root){
		if(root == null)
			return;
		print(root.left);
		System.out.println(root.val);
		print(root.right);
	}
	public int height(TreeNode root){
		int h = 0;
		if(root == null){
			return h;
		}
		else{
			h = 1 + Math.max(height(root.left), height(root.right));
		}
		return h;
	}
}

public class BalanceTree{
	public static void main(String[] args){
		Solution s = new Solution();
		TreeNode root = new TreeNode(1);
		TreeNode a = new TreeNode(2);
		TreeNode b = new TreeNode(2);
		root.left = a;
		root.right = b;

		TreeNode c = new TreeNode(3);
		TreeNode d = new TreeNode(3);
		TreeNode e = new TreeNode(3);
		TreeNode f = new TreeNode(3);
		a.left = c;
		a.right = d;
		b.left = e;
		b.right = f;

		TreeNode g = new TreeNode(4);
		TreeNode h = new TreeNode(4);
		TreeNode i = new TreeNode(4);
		TreeNode j = new TreeNode(4);
		TreeNode k = new TreeNode(4);
		TreeNode l = new TreeNode(4);
		c.left = g;
		c.right = h;
		d.left = i;
		d.right = j;
		e.left = k;
		e.right = l;

		TreeNode m = new TreeNode(5);
		TreeNode n = new TreeNode(5);
		g.left = m;
		g.right = n;

		System.out.println(s.isBalanced(root));
	}
}

File: python/leetcode/BalanceTree.py
'''
Leetcode problem No 110 Balanced Binary Tree
'''
class TreeNode(object):
	def __init__(self, x):
		self.val = x
		self.left = None
		self.right = None
	
class Solution(object):
	def height(self, root):
		if not root:
			return 0
		else: 
			return 1 + max(self.height(root.left), self.height(root.right))
	def isBalanced(self, root):
		if not root:
			return True
		if(abs(self.height(root.left) - self.height(root.right)) > 1):
			return False

		return self.isBalanced(root.left) and self.isBalanced(root.right)
		'''
		:type root: TreeNode
		:rtype: bool
		'''
def main():
    root = TreeNode(1)
    a = TreeNode(2)
    b = TreeNode(2)
    root.left = a
    root.right = b

    c = TreeNode(3)
    d = TreeNode(3)
    e = TreeNode(3)
    f = TreeNode(3)
    a.left = c
    a.right = d
    b.left = e
    b.right = f

    g = TreeNode(4)
    h = TreeNode(4)
    i = TreeNode(4)
    j = TreeNode(4)
    k = TreeNode(4)
    l = TreeNode(4)
    c.left = g
    c.right = h
    d.left = i
    d.right = j
    e.left = k
    e.right = l


    m = TreeNode(5)
    n = TreeNode(5)
    g.left = m
    g.right = n

    s = Solution()
    print(s.isBalanced(root))

main()
